streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -634,29 +634,32 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
       for {
         input       <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
         queueReader  = ZChannel.fromInput(input)
-        queue       <- Queue.bounded[ZIO[Env1, OutErr1, Either[OutDone, OutElem2]]](n)
-        _           <- scope.addFinalizer(queue.shutdown)
-        errorSignal <- Promise.make[OutErr1, Nothing]
+        outgoing    <- Queue.bounded[ZIO[Env1, Unit, Either[OutDone, OutElem2]]](n)
+        _           <- scope.addFinalizer(outgoing.shutdown)
+        errorSignal <- Promise.make[Unit, Nothing]
         permits     <- Semaphore.make(n.toLong)
+        failure     <- Ref.make[Cause[OutErr1]](Cause.empty)
         pull        <- (queueReader >>> self).toPullIn(scope)
         _ <- pull
                .foldCauseZIO(
-                 cause => queue.offer(ZIO.refailCause(cause)),
+                 cause => failure.update(_ && cause) *> outgoing.offer(ZIO.fail(())) *> ZIO.fail(()),
                  {
                    case Left(outDone) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> queue.offer(ZIO.succeed(Left(outDone)))
+                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(ZIO.succeed(Left(outDone)))
                    case Right(outElem) =>
                      for {
-                       p     <- Promise.make[OutErr1, OutElem2]
+                       p     <- Promise.make[Unit, OutElem2]
                        latch <- Promise.make[Nothing, Unit]
-                       _     <- queue.offer(p.await.map(Right(_)))
+                       _     <- outgoing.offer(p.await.map(Right(_)))
                        _ <- permits.withPermit {
                               latch.succeed(()) *>
                                 ZIO.uninterruptibleMask { restore =>
                                   restore(errorSignal.await) raceFirstAwait restore(f(outElem))
+                                    .tapErrorCause(cause => failure.update(_ && cause))
+                                    .mapError(_ => ())
                                 }
-                                  .tapErrorCause(errorSignal.failCause)
-                                  .intoPromise(p)
+                                  .flatMap(p.succeed)
+                                  .catchAll(_ => errorSignal.fail(()) *> p.fail(()))
```

Comment:
I think this whole block can be rewritten in a more optimized way as:

```scala
// In the companion object of Cause
val unit: Cause[Unit] = Cause.fail(())

// After raceFirstAwait
foldCauseZIO(
  cause => failure.update(_ && cause) *> errorSignal.refailCause(Cause.unit) *> p.refailCause(Cause.unit) *> Exit.unit,
  p.succeed
)
```

Reply:
update can only happen in the right branch, i don't want all those interruptions to be visible in the cause, otherwise all done